expression/builtin.go

```diff
@@ -173,9 +177,145 @@ func (b *baseBuiltinFunc) getArgs() []Expression {
 }
 
 func (b *baseBuiltinFunc) vecEval(input *chunk.Chunk, result *chunk.Column) error {
+	if b.self != nil && !b.vec {
+		return b.row2vec(input, result)
+	}
 	return errors.Errorf("baseBuiltinFunc.vecEval() should never be called, please contact the TiDB team for help")
 }
 
+func (b *baseBuiltinFunc) row2vec(input *chunk.Chunk, result *chunk.Column) error {
+	result.Reset()
+	it := chunk.NewIterator4Chunk(input)
+	sel := input.Sel()
+	if sel == nil {
+		switch b.tp.EvalType() {
+		case types.ETInt:
+			result.PreAllocInt64(input.NumEffectiveRows())
+			i64s := result.Int64s()
+			var isNull bool
+			var err error
+			row := it.Begin()
+			for i := range i64s {
+				i64s[i], isNull, err = b.self.evalInt(row)
+				if err != nil {
+					return err
+				}
+				if !isNull {
+					result.SetNull(i, false)
+				}
+				row = it.Next()
+			}
+		case types.ETReal:
+			result.PreAllocFloat64(input.NumEffectiveRows())
+			f64s := result.Float64s()
+			var isNull bool
+			var err error
+			row := it.Begin()
+			for i := range f64s {
+				f64s[i], isNull, err = b.self.evalReal(row)
+				if err != nil {
+					return err
+				}
+				if !isNull {
+					result.SetNull(i, false)
+				}
+				row = it.Next()
+			}
+		case types.ETDecimal:
+			result.PreAllocFloat64(input.NumEffectiveRows())
+			ds := result.Decimals()
+			var v *types.MyDecimal
+			var isNull bool
+			var err error
+			row := it.Begin()
+			for i := range ds {
+				v, isNull, err = b.self.evalDecimal(row)
+				if err != nil {
+					return err
+				}
+				if !isNull {
+					result.SetNull(i, false)
+					ds[i] = *v
+				}
+				row = it.Next()
+			}
+		default:
+			result.Reset()
+			for row := it.Begin(); row != it.End(); row = it.Next() {
+				if err := b.evalOneRow(row, result); err != nil {
+					return err
+				}
+			}
+		}
+	} else {
+		pos := 0
+		row := it.Begin()
+		for _, i := range sel {
+			for pos < i {
+				result.AppendNull()
```

Comment:
Why do we need to appendNull for the unselected rows?

Reply:
To align with row indexes in the `Sel` slice, we have to put placeholders to all unselected slots.